libdevcrypto/LibSnark.cpp

```diff
@@ -0,0 +1,212 @@
+/*
+ This file is part of cpp-ethereum.
+
+ cpp-ethereum is free software: you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ cpp-ethereum is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/**
+ * @file LibSnark.cpp
+ */
+
+#include <libdevcrypto/LibSnark.h>
+
+#include <algebra/curves/alt_bn128/alt_bn128_g1.hpp>
+#include <algebra/curves/alt_bn128/alt_bn128_g2.hpp>
+#include <algebra/curves/alt_bn128/alt_bn128_pairing.hpp>
+#include <algebra/curves/alt_bn128/alt_bn128_pp.hpp>
+
+#include <libdevcrypto/Exceptions.h>
+
+#include <libdevcore/CommonIO.h>
+#include <libdevcore/Log.h>
+
+using namespace std;
+using namespace dev;
+using namespace dev::crypto;
+
+namespace
+{
+
+void initLibSnark()
+{
+	// This is hackish, but we really want to use `static` variable for lock
+	// free thread-safe initialization.
+	static bool initialized = (libff::alt_bn128_pp::init_public_params(), true);
```

Comment:
Alternatively, you can wrap the function to add some logs to make sure it is called, and it is called once.